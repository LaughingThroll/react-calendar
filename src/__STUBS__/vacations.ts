import { VacationTypes } from '../types/model/vacation'
const vacations = [
  {
    startDate: '2021-01-15',
    endDate: '2021-01-18',
    type: VacationTypes.PAID,
  },
  {
    startDate: '2021-01-28',
    endDate: '2021-02-01',
    type: VacationTypes.PAID,
  },
  {
    startDate: '2021-02-12',
    endDate: '2021-02-15',
    type: VacationTypes.PAID,
  },
  {
    startDate: '2021-02-22',
    endDate: '2021-03-22',
    type: VacationTypes.UN_PAID,
  },
  {
    startDate: '2021-03-25',
    endDate: '2021-03-28',
    type: VacationTypes.PAID,
  },
]

export default vacations
