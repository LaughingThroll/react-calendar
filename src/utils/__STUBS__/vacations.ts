import { VacationType } from '../../types/model/vacation'
const vacations = [
  {
    startDate: '2021-01-15',
    endDate: '2021-01-18',
    type: VacationType.PAID,
  },
  {
    startDate: '2021-01-28',
    endDate: '2021-02-01',
    type: VacationType.PAID,
  },
  {
    startDate: '2021-02-12',
    endDate: '2021-02-15',
    type: VacationType.PAID,
  },
  {
    startDate: '2021-02-22',
    endDate: '2021-03-22',
    type: VacationType.UN_PAID,
  },
  {
    startDate: '2021-03-25',
    endDate: '2021-03-28',
    type: VacationType.PAID,
  },
]

export default vacations
