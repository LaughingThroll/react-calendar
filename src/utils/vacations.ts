import { reverseDate, normalizeUTCDate } from './date'

import { EVacationType, IVacation, IVacationDate } from '../types/model/vacation'

const checkVacation = (cellDate: Date, { startDate, endDate }: IVacationDate, separator: string = '.'): boolean => {
  return (
    cellDate >= new Date(normalizeUTCDate(reverseDate(startDate, separator))) &&
    cellDate <= new Date(normalizeUTCDate(reverseDate(endDate, separator)))
  )
}

const getCheckedDay = (date: Date, dateArr: string[]): boolean => {
  const [day, month] = dateArr.map(Number)
  return date.getDate() === day && date.getMonth() + 1 === month
}

const getBinaryNumber = (number: number): string => {
  return number < 10 ? `0${number}` : number.toString()
}

const getSplitedVacation = (vacation: IVacation, lastDay: number, separator: string = '.'): IVacation[] => {
  const { startDate, endDate, type } = vacation
  const [, startMonth, startYear] = startDate.split(separator).map(Number)
  const [, endMonth, endYear] = endDate.split(separator).map(Number)

  return [
    {
      startDate,
      endDate: [lastDay, getBinaryNumber(startMonth), startYear].join(separator),
      type,
    },
    {
      startDate: ['01', getBinaryNumber(endMonth), endYear].join(separator),
      endDate,
      type,
    },
  ]
}

export const getSplitVacations = (vacations: IVacation[], lastDay: number, separator: string = '.'): IVacation[] => {
  return vacations.flatMap((vacation) => {
    const { startDate, endDate, type } = vacation
    let [, startMonth] = startDate.split(separator).map(Number)
    const [, endMonth, endYear] = endDate.split(separator).map(Number)
    const diff = endMonth - startMonth

    if (diff === 1) return getSplitedVacation(vacation, lastDay)

    if (diff > 1) {
      const arrVacations = getSplitedVacation(vacation, lastDay)
      const getSplitedFullMonth = (month: number): IVacation => ({
        startDate: ['01', getBinaryNumber(month), endYear].join(separator),
        endDate: [lastDay, getBinaryNumber(month), endYear].join(separator),
        type,
      })

      while (startMonth !== endMonth - 1) {
        startMonth += 1
        arrVacations.push(getSplitedFullMonth(startMonth))
      }
      return arrVacations
    }
    return vacation
  })
}

export const getExsistingTypeVacation = (
  vacations: IVacation[],
  cellDate: Date,
  type: EVacationType.PAID | EVacationType.UN_PAID = EVacationType.PAID,
): boolean => {
  const arr = vacations.map(({ startDate, endDate, type }) =>
    checkVacation(cellDate, { startDate, endDate }) ? type : null,
  )
  return arr.some((el) => el === type)
}

// TODO: need think up new name and combine this function
export const isFirstDay = (vacations: IVacation[], date: Date, separator: string = '.'): boolean => {
  return vacations.map(({ startDate }) => getCheckedDay(date, startDate.split(separator))).some(Boolean)
}

export const isLastDay = (vacations: IVacation[], date: Date, separator: string = '.'): boolean => {
  return vacations.map(({ endDate }) => getCheckedDay(date, endDate.split(separator))).some(Boolean)
}

export const getDaysInVacation = ({ startDate, endDate }: IVacation, separator: string = '.') => {
  const [startDay] = startDate.split(separator).map(Number)
  const [endDay] = endDate.split(separator).map(Number)
  return endDay - startDay + 1
}

const getFilteredVacationsByMonth = (vacations: IVacation[], cellDate: Date, separator: string = '.'): IVacation[] => {
  return vacations.filter(({ startDate, endDate }) => {
    const [, startMonth, startYear] = startDate.split(separator).map(Number)
    const [, endMonth, endYear] = endDate.split(separator).map(Number)
    const currentMonth = cellDate.getMonth() + 1
    const currentYear = cellDate.getFullYear()
    return (
      (currentMonth === startMonth || currentMonth === endMonth) &&
      (currentYear === startYear || currentYear === endYear)
    )
  })
}

export const getSumVacationsDaysByMonth = (vacations: IVacation[], cellDate: Date): number => {
  const filteredArray = getFilteredVacationsByMonth(vacations, cellDate)
  console.log(filteredArray)
  return filteredArray.reduce((acc, vacation) => (acc += getDaysInVacation(vacation)), 0)
}

export const getSumVacationsDaysByDay = (vacations: IVacation[], cellDate: Date, separator: string = '.'): number => {
  const filteredVacations = getFilteredVacationsByMonth(vacations, cellDate)

  return filteredVacations.reduce((acc, { startDate, endDate }) => {
    const [startDay] = startDate.split(separator).map(Number)
    const [endDay] = endDate.split(separator).map(Number)
    return (acc += +(cellDate.getDate() >= startDay && cellDate.getDate() <= endDay))
  }, 0)
}
