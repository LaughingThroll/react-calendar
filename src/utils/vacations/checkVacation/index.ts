import { normalizeDate } from '../../date'
import { VacationDate } from '../../../types/model/vacation'

const checkVacation = (cellDate: Date, { startDate, endDate }: VacationDate): boolean => {
  return cellDate >= new Date(normalizeDate(startDate)) && cellDate <= new Date(normalizeDate(endDate))
}

export default checkVacation
