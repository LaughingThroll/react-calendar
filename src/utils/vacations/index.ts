export { default as checkVacation } from './checkVacation'
export { default as vacationIncludesVacation } from './vacationIncludesVacation'
export { default as getTypeVacation } from './getTypeVacation'
export { default as getSplitVacations } from './getSplitVacations'
export { default as getSumVacationsDaysByMonth } from './getSumVacationsDaysByMonth'
export { default as getSumVacationsDaysByDay } from './getSumVacationsDaysByDay'
export { default as isFirstDay } from './isFirstDay'
export { default as isLastDay } from './isLastDay'
