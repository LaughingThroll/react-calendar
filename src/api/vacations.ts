import { MAIN_URL } from './URLS'
import { makeRequest } from '../utils/'
import { VacationDTO, Vacation } from '../types/model/vacation'
import { FirebaseResponse } from '../types/model/firebase'

export const patchVacation = (submitVacation: VacationDTO): Promise<Vacation> => {
  const { currentTeamID, currentMemberID, ...vacation } = submitVacation
  return makeRequest<FirebaseResponse<Vacation>>(
    `${MAIN_URL}/teams/${currentTeamID}/members/${currentMemberID}/vacations.json`,
    { method: 'PATCH', body: JSON.stringify({ [Date.now()]: vacation }) }
  ).then((vacation) => Object.values(vacation)[0])
}
