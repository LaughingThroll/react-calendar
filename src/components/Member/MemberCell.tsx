import React from 'react'
import classnames from 'classnames'

import { useCalucateWidth } from '../../hooks'
import { isWeekend, isEqualDate, getCountDays, reverseDate } from '../../utils/date'
import { getTypeVacation, isFirstDay, isLastDay } from '../../utils/vacations'
import { VacationType, IVacation } from '../../types/model/vacation'

interface IMemberCell {
  date: Date
  vacations: IVacation[]
}

const MemberCell: React.FC<IMemberCell> = ({ date, vacations }) => {
  const vacationElement = useCalucateWidth<Date, HTMLDivElement>(date)
  const isPaid = getTypeVacation(vacations, date, VacationType.PAID)
  const isUnPaid = getTypeVacation(vacations, date, VacationType.UN_PAID)
  const isStartDay = isFirstDay(vacations, date)
  const isEndDay = isLastDay(vacations, date)

  const currentStartVacation: IVacation | undefined = vacations.find(({ startDate }) => {
    return isEqualDate(date, startDate.split('.'))
  })
  const countDayInVacation =
    currentStartVacation &&
    getCountDays(reverseDate(currentStartVacation.startDate), reverseDate(currentStartVacation.endDate)) + 1

  return (
    <td
      className={classnames({
        'calendar-body__cell': true,
        'cell-gray': isWeekend(date),
      })}
    >
      {(isUnPaid || isPaid) && (
        <div
          ref={vacationElement.elementRef}
          className={classnames({
            'cell-vacations': true,
            'cell-vacations--paid': isPaid,
            'cell-vacations--unpaid': isUnPaid,
            'cell-vacations--start-vac': isStartDay,
            'cell-vacations--end-vac': isEndDay,
            'cell-vacations--unpaid-start-vac': isStartDay && isUnPaid,
            'cell-vacations--unpaid-end-vac': isEndDay && isUnPaid,
          })}
        >
          {countDayInVacation && (
            <span style={{ width: vacationElement.width * countDayInVacation }} className="cell-vacations__text">
              Pd
            </span>
          )}
        </div>
      )}
    </td>
  )
}

export default React.memo(MemberCell)
