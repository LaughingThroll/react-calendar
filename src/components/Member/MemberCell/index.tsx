import React from 'react'
import classnames from 'classnames'

import { Cell } from '../../common'

import { useCalucateWidth } from '../../../hooks'
import { isEqualDate, getCountDays } from '../../../utils/date'
import { getTypeVacation, isFirstDay, isLastDay } from '../../../utils/vacations'
import { VacationTypes, Vacation } from '../../../types/model/vacation'

export interface MemberCellProps {
  date: Date
  vacations: Vacation[]
}

const MemberCell: React.FC<MemberCellProps> = ({ date, vacations }) => {
  const vacationElement = useCalucateWidth<Date, HTMLDivElement>(date)
  const isPaid = getTypeVacation(vacations, date, VacationTypes.PAID)
  const isUnPaid = getTypeVacation(vacations, date, VacationTypes.UN_PAID)
  const isStartDay = isFirstDay(vacations, date)
  const isEndDay = isLastDay(vacations, date)

  const currentStartVacation: Vacation | undefined = vacations.find(({ startDate }) => {
    return isEqualDate(date, startDate.split('-'))
  })
  const countDayInVacation =
    currentStartVacation && getCountDays(currentStartVacation.startDate, currentStartVacation.endDate) + 1

  return (
    <Cell cellDate={date} classNames={['calendar-body__cell']}>
      {(isUnPaid || isPaid) && (
        <div
          ref={vacationElement.elementRef}
          className={classnames({
            'cell-vacations': true,
            'cell-vacations--paid': isPaid,
            'cell-vacations--unpaid': isUnPaid,
            'cell-vacations--start-vac': isStartDay,
            'cell-vacations--end-vac': isEndDay,
            'cell-vacations--unpaid-start-vac': isStartDay && isUnPaid,
            'cell-vacations--unpaid-end-vac': isEndDay && isUnPaid,
          })}
        >
          {countDayInVacation && (
            <span style={{ width: vacationElement.width * countDayInVacation }} className="cell-vacations__text">
              Pd
            </span>
          )}
        </div>
      )}
    </Cell>
  )
}

export default React.memo(MemberCell)
