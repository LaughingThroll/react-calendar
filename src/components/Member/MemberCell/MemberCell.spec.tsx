import { shallow } from 'enzyme'
import MemberCell, { MemberCellProps } from './'
import vacations from '../../../__STUBS__/vacations'

const setUp = (props: MemberCellProps) => shallow(<MemberCell {...props} />)

describe('MemberCell render', () => {
  it('default', () => {
    const props = { date: new Date('2021-08-05'), vacations }
    const tree = setUp(props)

    expect(tree).toMatchSnapshot()
  })

  it('with vacation type Paid', () => {
    const props = { date: new Date('2021-01-15'), vacations }
    const tree = setUp(props)

    expect(tree).toMatchSnapshot()
  })

  it('with vacation type UnPaid', () => {
    const props = { date: new Date('2021-03-20'), vacations }
    const tree = setUp(props)

    expect(tree).toMatchSnapshot()
  })
})
