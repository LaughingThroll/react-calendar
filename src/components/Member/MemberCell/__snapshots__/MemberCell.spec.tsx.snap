// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberCell render default 1`] = `
<Memo(Cell)
  cellDate={2021-08-05T00:00:00.000Z}
  classNames={
    Array [
      "calendar-body__cell",
    ]
  }
/>
`;

exports[`MemberCell render with vacation type Paid 1`] = `
<Memo(Cell)
  cellDate={2021-01-15T00:00:00.000Z}
  classNames={
    Array [
      "calendar-body__cell",
    ]
  }
>
  <div
    className="cell-vacations cell-vacations--paid cell-vacations--start-vac"
  >
    <span
      className="cell-vacations__text"
      style={
        Object {
          "width": 0,
        }
      }
    >
      Pd
    </span>
  </div>
</Memo(Cell)>
`;

exports[`MemberCell render with vacation type UnPaid 1`] = `
<Memo(Cell)
  cellDate={2021-03-20T00:00:00.000Z}
  classNames={
    Array [
      "calendar-body__cell",
    ]
  }
>
  <div
    className="cell-vacations cell-vacations--unpaid"
  />
</Memo(Cell)>
`;
